quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,125 @@
+//! Pacing of packet transmissions.
+
+use std::time::{Duration, Instant};
+
+use tracing::{debug, warn};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate slightly faster
+/// than one contestion window per RTT.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    prev: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64, now: Instant) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            prev: now,
+        }
+    }
+
+    /// Record that a packet has been transmitted.
+    pub fn on_transmit(&mut self, packet_length: u16) {
+        self.tokens = self.tokens.saturating_sub(packet_length.into())
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns `None`. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// The 5/4 ratio used here comes from the suggestion that N = 1.25 in the draft IETF RFC for
+    /// QUIC.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
+        now: Instant,
+    ) -> Option<Instant> {
+        debug_assert_ne!(
+            window, 0,
+            "zero-sized congestion control window is nonsense"
+        );
+
+        // if we can already send a packet, there is no need for delay
+        if self.tokens > mtu.into() {
+            return None;
+        }
+
+        // we disable pacing for extremely large windows
+        if window > u32::max_value().into() {
+            debug!("window too large, disabling pacing");
```

Comment:
I feel like this might get really spammy under certain circumstances, and repeating it probably isn't all that useful. Maybe just omit?